parlai/core/distributed_utils.py

```diff
@@ -0,0 +1,201 @@
+#!/usr/bin/env python3
+
+# Copyright (c) 2017-present, Facebook, Inc.
+# All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree. An additional grant
+# of patent rights can be found in the PATENTS file in the same directory.
+
+
+"""
+Useful utilities for training in distributed mode.
+"""
+
+import builtins
+import pickle
+try:
+    import torch.version
+    import torch.distributed as dist
+    TORCH_AVAILABLE = True
+except ImportError:
+    TORCH_AVAILABLE = False
+
+
+def validate_params(opt):
+    """
+    Ensure sane combinations of command line parameters for distributed training.
+
+    Raises exceptions if anything is wrong, otherwise returns None.
+    """
+    if torch.version.__version__.startswith('0.'):
+        raise ImportError(
+            "Please upgrade to PyTorch >=1.0; "
+            "visit https://pytorch.org for instructions."
+        )
+
+    if opt.get('no_cuda', False):
+        raise ValueError(
+            'Distributed mode only makes sense when using GPUs.'
+        )
+
+    if opt.get('numthreads', 1) != 1:
+        raise ValueError(
+            '--numthreads must be 1 for distributed training.'
+        )
+
+    if 'train:stream' in opt['datatype'] or 'ordered' in opt['datatype']:
+        raise ValueError(
+            "You should not combine ordered streaming with distributed training "
+            "because all workers will have exactly the same minibatches, "
+            "defeating the purpose."
+        )
+
+
+def is_distributed():
+    """
+    Returns True if we are in distributed mode.
+    """
+    return TORCH_AVAILABLE and dist.is_initialized()
+
+
+def num_workers():
+    """
+    Get the total number of workers.
+    """
+    if not is_distributed():
+        return 1
+    else:
+        return dist.get_world_size()
+
+
+def is_primary_worker():
+    """
+    Returns False if we are a secondary worker. Returns True if we are either
+    (1) not in distributed mode (2) or are the primary (rank 0) worker.
+    """
+    return not is_distributed() or dist.get_rank() == 0
+
+
+def override_print(suppress=False, prefix=None):
+    """
+    Overrides the builtin print, to either mute or annotate the output of
+    individual workers. Mutes if suppress is True, and this is not the
```

Comment:
nit: does it suppress if mute is true AND this is not the primary/only worker? or dies it suppress if mute is true OR this is not the primary/only worker?

Reply:
I will update the docstring correspondingly.